geometry/include/pcl/geometry/mesh_indices.h

```diff
@@ -38,314 +38,54 @@
-  VertexIndex() : index_(-1) {}
-
-  /** \brief Constructor.
-   * \param[in] index The integer index.
-   */
-  explicit VertexIndex(const int index) : index_(index) {}
-
-  /** \brief Returns true if the index is valid. */
-  inline bool
-  isValid() const
-  {
-    return (index_ >= 0);
-  }
-
-  /** \brief Invalidate the index. */
-  inline void
-  invalidate()
-  {
-    index_ = -1;
-  }
-
-  /** \brief Get the index. */
-  inline int
-  get() const
-  {
-    return (index_);
-  }
+template <class MeshIndexT> 
+class MeshIndex;
 
-  /** \brief Set the index. */
-  inline void
-  set(const int index)
-  {
-    index_ = index;
-  }
+template <class MeshIndexT> 
+std::istream& 
+operator>>(std::istream& is, pcl::detail::MeshIndex<MeshIndexT>&);
 
-  /** \brief Comparison operators (with boost::operators): < > <= >= */
-  inline bool
-  operator<(const Self& other) const
-  {
-    return (this->get() < other.get());
-  }
-
-  /** \brief Comparison operators (with boost::operators): == != */
-  inline bool
-  operator==(const Self& other) const
-  {
-    return (this->get() == other.get());
-  }
-
-  /** \brief Increment operators (with boost::operators): ++ (pre and post) */
-  inline Self&
-  operator++()
-  {
-    ++index_;
-    return (*this);
-  }
-
-  /** \brief Decrement operators (with boost::operators): \-\- (pre and post) */
-  inline Self&
-  operator--()
-  {
-    --index_;
-    return (*this);
-  }
-
-  /** \brief Addition operators (with boost::operators): + += */
-  inline Self&
-  operator+=(const Self& other)
-  {
-    index_ += other.get();
-    return (*this);
-  }
-
-  /** \brief Subtraction operators (with boost::operators): - -= */
-  inline Self&
-  operator-=(const Self& other)
-  {
-    index_ -= other.get();
-    return (*this);
-  }
-
-private:
-  /** \brief Stored index. */
-  int index_;
-
-  friend std::istream&
-  operator>>(std::istream& is, pcl::geometry::VertexIndex& index);
-};
-
-/** \brief ostream operator. */
-inline std::ostream&
-operator<<(std::ostream& os, const pcl::geometry::VertexIndex& index)
-{
-  return (os << index.get());
-}
-
-/** \brief istream operator. */
-inline std::istream&
-operator>>(std::istream& is, pcl::geometry::VertexIndex& index)
-{
-  return (is >> index.index_);
-}
-
-} // End namespace geometry
-} // End namespace pcl
-
-////////////////////////////////////////////////////////////////////////////////
-// HalfEdgeIndex
-////////////////////////////////////////////////////////////////////////////////
-
-namespace pcl {
-namespace geometry {
-/** \brief Index used to access elements in the half-edge mesh. It is basically just a
- * wrapper around an integer with a few added methods.
- * \author Martin Saelzle
- * \ingroup geometry
- */
-class HalfEdgeIndex
+template <class MeshIndexT>
+class MeshIndex
 : boost::totally_ordered<
-      pcl::geometry::HalfEdgeIndex // < > <= >= == !=
+      pcl::detail::MeshIndex<MeshIndexT> // < > <= >= == !=
```

Comment:
Could you see if we can skip `pcl::detail::` (and move the comma to the previous line)